docs/source/argument_check.rst

```diff
@@ -0,0 +1,636 @@
+
+
+Device Memory Size Queries
+==========================
+
+- When ``handle->is_device_memory_size_query()`` is true, the call is not a normal call, but it is a device memory size query.
+
+- No logging should be performed during device memory size queries.
+
+- If the rocBLAS kernel requires no temporary device memory, the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle)`` can be called after checking that ``handle != nullptr``.
+
+- If the rocBLAS kernel requires temporary device memory, then it should be set, and the kernel returned, by calling ``return handle->set_optimal_device_memory_size(size...)``, where ``size...`` is a list of one or more sizes for different sub-problems. The sizes are rounded up and added.
+
+Logging
+-------
+
+- There is logging before a quick-return-success or error-return, except:
+
+  - when ``handle == nullptr``, return ``rocblas_status_invalid_handle`` 
+  - when ``handle->is_device_memory_size_query()`` returns ``true``
+
+- Vectors and matrices are logged with their addresses, and are always on device memory.
+
+- Scalar values in device memory are logged as their addresses. Scalar values in host memory are logged as their values, with a ``nullptr`` logged as ``NaN`` (``std::numeric_limits<T>::quiet_NaN()``).
+
+rocBLAS control flow:
+=====================
+
+1. If ``handle == nullptr``, return ``rocblas_status_invalid_handle``.
+
+2. If the function does not require temporary device memory, call the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle);``.
+
+3. If the function requires temporary device memory, and ``handle->is_device_memory_size_query()`` is ``true``, then validate any pointers and arguments required to determine the optimal size of temporary device memory, returning ``rocblas_status_invalid_pointer`` or ``rocblas_status_invalid_size`` if the arguments are invalid, and otherwise ``return handle->set_optimal_device_memory_size(size...);``, where ``size...`` is a list of one or more sizes of temporary buffers, which are allocated with ``handle->device_malloc(size...)`` later.
+
+4. Perform logging if enabled, taking care not to dereference ``nullptr`` arguments.
+
+5. Check for invalid sizes. Return ``rocblas_status_invalid_size`` if size arguments are invalid.
+
+3. Check for unsupported enum value. Return ``rocblas_status_invalid_value`` if enum value is invalid.
+
+6. Return ``rocblas_status_invalid_pointer`` if any pointers used to determine quick return conditions are NULL.
+
```

Comment:
I think it is not just quick return condition testing that are tested.   Could it be said that any pointer required for calculation is NULL. (i.e. a pointer that is not ignored based on other arguments, for example N == 0 means ignore input pointers)

Reply:
I think after correcting the numbering is should be more clear.

AFAIK the Fortran philosophy is that if you can quick_return_success because N == 0 you do not need to check if pointers are NULL pointers. 

In the axpy example, there is a rocblas_status_success return if  alpha !=0 && *alpha == 0 before the check to return rocblas_status_invalid_pointer  if !x  || !y